app/src/org/commcare/dalvik/activities/EntityMapActivity.java

```diff
@@ -1,288 +1,186 @@
-        //The overlay crashes out if you try to draw it and it's empty,
-        //so only add it if we found something.
-        if(mEntityOverlay.size() > 0) {
-            map.getOverlays().add(mEntityOverlay);
-        }
-        map.getController().setZoom(18);
-        map.setClickable(true);
-        map.setEnabled(true);
-        Log.d(TAG, "Done loading");
+        Log.d(TAG, "Loaded. " + entityLocations.size() +" addresses discovered, " + bogusAddresses + " could not be located");
     }
-    
-    private String getLicenseKey() {
-        //If there's a defined debug key in the local environment, use that.
-        int debugId = this.getResources().getIdentifier("maps_api_key_debug","string", this.getPackageName());
-        if(debugId == 0) { 
-            return BuildConfig.MAPS_API_KEY;
+
+    private LatLng getLatLngFromAddress(String address) {
+        LatLng location = null;
+        try {
+            GeoPointData data = new GeoPointData().cast(new UncastData(address));
+            if(data != null) {
+                location = new LatLng(data.getLatitude(), data.getLongitude());
+            }
+        } catch(Exception ex) {
         }
-        return this.getString(debugId);
-    }
 
-    private EvaluationContext entityContext;
+        // Geocaching code removed
 
-    private EvaluationContext getEC() {
-        if(entityContext == null) {
-            entityContext = session.getEvaluationContext(getInstanceInit());
-        }
-        return entityContext;
-    }
-    
-    private AndroidInstanceInitializer getInstanceInit() {
-        return new AndroidInstanceInitializer(session);
+        return location;
     }
 
     @Override
-    protected void onStop() {
-        super.onStop();
-        mMyLocationOverlay.disableCompass(); 
-        mMyLocationOverlay.disableMyLocation();
+    public void onMapReady(final GoogleMap map) {
+        mMap = map;
+
+        // Find bounding region of markers
+        LatLngBounds.Builder builder = new LatLngBounds.Builder();
+        for (Pair<Entity<TreeReference>, LatLng> entityLocation: entityLocations) {
+            Marker marker = map.addMarker(new MarkerOptions()
+                    .position(entityLocation.second)
+                    .title(entityLocation.first.getFieldString(0))
+                    .snippet(entityLocation.first.getFieldString(1)));
+            markerReferences.put(marker, entityLocation.first.getElement());
+            builder.include(entityLocation.second);
+        }
+        final LatLngBounds bounds = builder.build();
+
+        // Move camera to be include all markers
+        // TODO: does this work for 0 or 1 marker?
+        map.setOnMapLoadedCallback(new GoogleMap.OnMapLoadedCallback() {
+            @Override
+            public void onMapLoaded() {
+                map.animateCamera(CameraUpdateFactory.newLatLngBounds(bounds, MAP_PADDING));
+            }
+        });
+
+        map.setOnInfoWindowClickListener(this);

```

Comment:
I tried refactoring it to re-register it in onPause, but it ended up getting messy, because to avoid doing the whole bounds calculation twice you have to make new boolean flags. I'm not sure it's worth it. OnMapReady gets called before loading, so the only way the callback might not be registered is if onMapReady is actually interrupted, and it doesn't do any blocking stuff as far as I can tell.